netdata/netdata
# Modify PID monitoring (ebpf.plugin)

##### Summary
This PR is bringing new libraries and it is also reducing memory usage for ebpf.plugin bringing a new way to monitor PID. The PID monitoring will also have more PRs bringing changes.

##### Test Plan

1. Compile master branch
2. Change your `ebpf.d.conf` to:
```conf
[ebpf programs]
    cachestat = yes
    dcstat = yes
```
3. Run `bpftool map show` and take a look in tables `tbl_pid_stats`, `cstat_pid`, and `dcstat_pid`, they will have `max_entries` with value `32768`.
4. Stop netdata and compile this branch
5. Start netdata again and run `bpftool map show`, this time the tables listed in `3.` will have `max_entries` with value `10992` and of course the memory usage is reduced.
6. Run `grep libbpf /var/log/netdata/error.log`, you should not have any error for kernels `4.11`or newer, for ancient kernels this PR is not touching them.

##### Additional Information
The code added with this PR was intially tested using [CO-RE](https://github.com/netdata/ebpf-co-re/pull/27) codes and also legacy codes:

| Distribution | kernel | real parent | parent | all |
|-----------------|---------|-----------------|----------|-----|
|Slackware current | 5.18.16 | [slackware_5_18_16_pid0.txt](https://github.com/netdata/netdata/files/9296040/slackware_5_18_16_pid0.txt) | [slackware_5_18_16_pid1.txt](https://github.com/netdata/netdata/files/9296041/slackware_5_18_16_pid1.txt) | [slackware_5_18_16_pid2.txt](https://github.com/netdata/netdata/files/9296042/slackware_5_18_16_pid2.txt) |
| Arch Linux | 5.18.16-arch-1 |  [arch_5_18_16_pid0.txt](https://github.com/netdata/netdata/files/9296049/arch_5_18_16_pid0.txt) | [arch_5_18_16_pid1.txt](https://github.com/netdata/netdata/files/9296050/arch_5_18_16_pid1.txt) | [arch_5_18_16_pid2.txt](https://github.com/netdata/netdata/files/9296051/arch_5_18_16_pid2.txt) |
| Ubuntu 22.04 | 5.15.0-33-generic |  [ubuntu_5_15_pid0.txt](https://github.com/netdata/netdata/files/9296052/ubuntu_5_15_pid0.txt) | [ubuntu_5_15_pid1.txt](https://github.com/netdata/netdata/files/9296053/ubuntu_5_15_pid1.txt) | [ubuntu_5_15_pid2.txt](https://github.com/netdata/netdata/files/9296054/ubuntu_5_15_pid2.txt) |
| Ubuntu 20.04 | 5.11.0-38-generic |  [ubuntu_5_11_pid0.txt](https://github.com/netdata/netdata/files/9296062/ubuntu_5_11_pid0.txt) |  [ubuntu_5_11_pid1.txt](https://github.com/netdata/netdata/files/9296063/ubuntu_5_11_pid1.txt) | [ubuntu_5_11_pid2.txt](https://github.com/netdata/netdata/files/9296064/ubuntu_5_11_pid2.txt) | 
| Alma 9.0 | 5.14.0-70.17.1.el9_0 | [alma_5_14_pid0.txt](https://github.com/netdata/netdata/files/9296068/alma_5_14_pid0.txt) | [alma_5_14_pid1.txt](https://github.com/netdata/netdata/files/9296069/alma_5_14_pid1.txt) | [alma_5_14_pid2.txt](https://github.com/netdata/netdata/files/9296070/alma_5_14_pid2.txt) | 
| Alma 8.6 |  4.18.0-372.19.1.el8_6 | [alma_4_18_pid0.txt](https://github.com/netdata/netdata/files/9296076/alma_4_18_pid0.txt) |    [alma_4_18_pid1.txt](https://github.com/netdata/netdata/files/9296077/alma_4_18_pid1.txt) | [alma_4_18_pid2.txt](https://github.com/netdata/netdata/files/9296078/alma_4_18_pid2.txt) |

And for all tests we observed the expected result that less PIDs are stored, so we can reduce memory and cpu usage.

This PR was tested on:

| Distribution | kernel | bpftool output | threads running |
|-----------------|----------|--------------------|----------------------|
| Slackware current | 5.18.16 | [slackware_5_18_bpftool.txt](https://github.com/netdata/netdata/files/9303373/slackware_5_18_bpftool.txt) | [slackware_5_18_threads.txt](https://github.com/netdata/netdata/files/9303375/slackware_5_18_threads.txt) |
| Arch | 5.18.16-arch1 | [arch_5_18_bpftool.txt](https://github.com/netdata/netdata/files/9303429/arch_5_18_bpftool.txt) | [arch_5_18_threads.txt](https://github.com/netdata/netdata/files/9303430/arch_5_18_threads.txt) |
| Ubuntu 22.04 | 5.15.0-33-generic | [ubuntu_5_15_bpftool.txt](https://github.com/netdata/netdata/files/9303434/ubuntu_5_15_bpftool.txt) | [ubuntu_5_15_threads.txt](https://github.com/netdata/netdata/files/9303435/ubuntu_5_15_threads.txt) 
| Alma 9.0 | 5.14.0-70.17.1.el9_0 |  [alma_5_14_bpftool.txt](https://github.com/netdata/netdata/files/9303440/alma_5_14_bpftool.txt) | [alma_5_14_threads.txt](https://github.com/netdata/netdata/files/9303441/alma_5_14_threads.txt) |
| Ubuntu 20.04 | 5.11.0-38-generic | [ubuntu_5_11_threads.txt](https://github.com/netdata/netdata/files/9303457/ubuntu_5_11_threads.txt) | [alma8_4_18_bpftool.txt](https://github.com/netdata/netdata/files/9303458/alma8_4_18_bpftool.txt) |
| Alma 8.6 | 4.18.0-372.19.1.el8_6 | [alma8_4_18_bpftool.txt](https://github.com/netdata/netdata/files/9304111/alma8_4_18_bpftool.txt) | [alma8_4_18_threads.txt](https://github.com/netdata/netdata/files/9303462/alma8_4_18_threads.txt) |
| Ubuntu 18.04 | 4.15.0-180 |  [ubuntu_4_15_bpftool.txt](https://github.com/netdata/netdata/files/9303465/ubuntu_4_15_bpftool.txt) | [ubuntu_4_15_threads.txt](https://github.com/netdata/netdata/files/9303466/ubuntu_4_15_threads.txt) |

<details> <summary>For users: How does this change affect me?</summary>
Describe the PR affects users: 
- Which area of Netdata is affected by the change? ebpf.plugin
- Can they see the change or is it an under the hood? If they can see it, where? ebpf.plugin will use less memory and cpu.
- How is the user impacted by the change? A collector using less memory and cpu.
- What are there any benefits of the change? A better monitoring
</details>
